File: BaekJoon_16236.py
# Baekjoon 16236
# https://www.acmicpc.net/problem/16236

from sys import stdin
from collections import deque
import heapq

rd = stdin.readline

def bfs():
    global answer
    tmp = 0
    while q:
        tmp += 1
        next_steps = []
        step = len(q)
        for _ in " " * step:
            x, y = q.popleft()
            for mv_idx in range(4):
                nx, ny = x + dx[mv_idx], y + dy[mv_idx]
                if 0 <= nx < N and 0 <= ny < N and not visited[nx][ny] and space[nx][ny] <= shark_size: # 상어와 크기가 같거나 작은 물고기가 있는 칸으로만 이동 가능
                    if 0 < space[nx][ny] < shark_size:
                        next_steps.append((nx, ny))
                    visited[nx][ny] = True
                    q.append((nx, ny))
        heapq.heapify(next_steps)
        for nx, ny in next_steps:
            answer += tmp
            need_to_growth[shark_size] -= 1 # 한마리를 먹음
            fish_count[space[nx][ny]] -= 1
            space[shark_coord[0]][shark_coord[1]] = 0   # 원래 상어의 위치는 빈 칸으로 변경
            space[nx][ny] = 9   # 물고기를 먹은 위치로 상어 이동
            q.clear()   # 먹은 위치부터 bfs재시작 해야하기에 초기화 후 현재위치만 저장
            q.append((nx, ny))
            return [nx, ny]

dx = [-1, 0, 0, 1]  # 위, 왼쪽, 오른쪽, 아래
dy = [0, -1, 1, 0]
need_to_growth = [0, 0, 2, 3, 4, 5, 6, 400]
fish_count = [0] * 7
q = deque()

N = int(rd())
space = []
shark_coord = [0, 0]
for r in range(N):
    input_row = list(map(int, rd().split()))
    space.append(input_row)
    for c, v in enumerate(input_row):
        if v:
            if v == 9:
                q.append((r, c))
                shark_coord = [r, c]
            else:
                fish_count[v] += 1

answer = 0
shark_size = 2
while sum(fish_count[:shark_size]) and q: # 먹을 수 있는 물고기 (상어 크기보다 작은 물고기)가 존재하는 동안 계속 실행
    visited = [[False] * N for _ in " " * N]
    shark_coord = bfs() # 물고기를 먹을 수 있는 위치에 도달한 경우 bfs종료되므로 해당 위치로 상어의 좌표를 옮겨줌
    if not need_to_growth[shark_size]:
        shark_size += 1

print(answer)

File: BaekJoon_16236.java
// Baekjoon 16236
// https://www.acmicpc.net/problem/16236

import java.io.*;
import java.util.*;
public class Main {
    static int[] dx = new int[]{-1, 1, 0, 0};
    static int[] dy = new int[]{0, 0, 1, -1};
    static int answer = 0;

    static int[] chooseNextPos(List<int[]> positions) {
        int[][] arr = new int[positions.size()][2];    // 행, 열을 기준으로 정렬하기 위해서 리스트를 2중 배열로 변환
        for (int idx = 0; idx < positions.size(); idx++) {
            arr[idx] = positions.get(idx);
        }
        Arrays.sort(arr, new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                if(o1[0] == o2[0]) {
                    return o1[1] - o2[1];
                }else {
                    return o1[0] - o2[0];
                }
            }
        });
        return arr[0];
    }

    static int[] bfs(int n, int[][] map, int shark_size, int[] shark_pos) {
        boolean[][] visited = new boolean[n][n];
        int distance = 0;
        Deque<int[]> deque = new LinkedList<>();
        deque.offer(shark_pos);
        while (!deque.isEmpty()) {
            distance++;
            List<int[]> canEatFish = new ArrayList<>();
            int step = deque.size();
            for (int s=0; s<step; s++) {
                int[] pos = deque.poll();
                for (int mv_idx=0; mv_idx<4; mv_idx++) {
                    int nx = pos[0] + dx[mv_idx];
                    int ny = pos[1] + dy[mv_idx];
                    if ((0 <= nx && nx < n) && (0 <= ny && ny < n) && !visited[nx][ny] && map[nx][ny] <= shark_size) {
                        visited[nx][ny] = true;
                        deque.offer(new int[]{nx, ny});
                        if (0 < map[nx][ny] && map[nx][ny] < shark_size)
                            canEatFish.add(new int[]{nx, ny});
                    }
                }
            }

            if (!canEatFish.isEmpty()) {
                answer += distance;
                return chooseNextPos(canEatFish);
            }
        }
        return new int[]{-1, -1};
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[][] map = new int[N][N];

        int[] shark_pos = new int[2];
        for (int i=0; i<N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j=0; j<N; j++) {
                 map[i][j] = Integer.parseInt(st.nextToken());
                 if (map[i][j] == 9)
                     shark_pos = new int[]{i, j};
            }
        }
        br.close();
        
        map[shark_pos[0]][shark_pos[1]] = 0;

        int shark_size = 2, exp = 0;
        while (true) {
            shark_pos = bfs(N, map, shark_size, shark_pos);
            if (Arrays.equals(shark_pos, new int[]{-1, -1})) {
                break;
            }
            exp++;
            map[shark_pos[0]][shark_pos[1]] = 0;
            if (shark_size == exp) {
                shark_size++;
                exp = 0;
            }
        }
        System.out.println(answer);
    }
}
